Filterx scope optimizations

bazsi opened the pull request:
This goes on top of #467 and brings significant performance improvements for FilterXScope

MrAnno commented on lib/filterx/filterx-scope.h:
```diff
@@ -43,11 +43,22 @@ typedef struct _FilterXScope FilterXScope;
+    union
+    {
+      FilterXVariable *separate;
+      FilterXVariable coupled[0];
```
Does it really worth it to save an indirection and an initial malloc to inline variables into the scope?
If it's really worth the code complexity and want to keep it, can we rename `coupled` to `inlined`, `separate` to `allocated`, and `coupled_variables` to `inlined_variables`?

bazsi replied:
We have quite a few FilterXScope instances, as each and every filterx block creates one. And we can argue that all of our expressions operate on variables, which are stored in these slots. IIRC in the case of our palo alto pipeline, there were 37 FilterXScope instances which generated 37 allocations for FilterXScope and then the same amount for the GPtrArray instance and then the pointer array within that GPtrArray instance.

I don't remember how much this was worth in performance improvements, but I measured all of these changes and dropped those which didn't have a measurable impact.

I went back to some of my notes on Slack. I've found a relatively old profile where filterx_scope_unref() (with the associated cost in _freeing_ of the variables array) was responsible for 10.9% of the cycles while the one with the new layout it was 3.3%

Another comparison, all FilterXScope related operations are 7.2% on a new profile, and 33.7% on the old one.

bazsi replied:
wrt naming, I kind of liked those as I worked with them (even though I do change names while working on a specific code, until I feel they are descriptive enough. I guess this is a matter of style. I don't mind renaming them to what you recommended (might be easier to understand for some), but I think these are good enough for merging, especially as this seems to fix a blocker bug and improves performance a lot.

MrAnno commented on lib/filterx/filterx-scope.c:
```diff
@@ -356,12 +420,21 @@ _free(FilterXScope *self)
-  if (variables_max < self->variables->len)
-    filterx_scope_variables_max = self->variables->len;
+  if (variables_max < self->variables.len)
+    filterx_scope_variables_max = self->variables.len;
```
I think we should put a hard-coded upper limit to this variable.

If someone loads a configuration with millions of variables, not even a reload will help reduce the memory usage.

bazsi replied:
I am not sure how that could happen, we don't have dynamic allocation of variables, we do not have loops, etc. The only way to create 1M variables is to create a configuration with 1M lines, each defining a variable.

With that said, the usual stack size is 8M, a FilterXVariable is 16 bytes, but we need stack space for the actual function calls, and as we just discussed, FilterXScopes are embedded on the stack. What if we just limited the number of maximum variables any FilterXScope can support to 1024? Or maybe 4096? I don't think anyone would need more, or if they do, we can always bump that higher.

That should limit the stack use to 64kB

MrAnno replied:
Maybe I missed something, but I'm only talking about the maximum number of variables that we allocate on the stack. Heap-allocated variables could remain infinite in my opinion.

My suggested would be to rename and calculate `filterx_scope_variables_max` like this:
`filterx_scope_variables_max_inlined = MIN(self->variables.len, 1024)`

MrAnno commented on lib/filterx/filterx-pipe.c:
```diff
@@ -66,9 +66,22 @@ log_filterx_pipe_queue(LogPipe *s, LogMessage *msg, const LogPathOptions *path_o
+
+  if (!scope)
+    {
+      gsize alloc_size = filterx_scope_get_alloc_size();
```
Now that we are on the stack, we should definitely put an upper limit on the maximum allocated variables count.

MrAnno reviewed (commented):
Please revert ` f917c67d02d6a60883ed7da7915bdaf37338a998` in this PR.

bazsi commented:
Rebased the branch, added the revert and reimplemented a small optimization that was part of the revert. I did not change any other things yet. I looked into adding the limit, but unfortunately most things that take care of variables basically assume we have an unlimited supply of them, errors are not handled. This is something I wouldn't start at 9pm.

bazsi reviewed (commented):
THe new patches look good to me. Thanks @MrAnno
